Add UI Toolkit matching control bindings UI

lyndon-unity opened the pull request:
### Description

Implemented UI TK version of the matching control bindings recently added to the IMGUI editor
https://github.com/Unity-Technologies/InputSystem/pull/1672

Example of it working in the project wide input actions. This is the same UX used in the individual assets.
![image](https://github.com/Unity-Technologies/InputSystem/assets/33493311/2f4ec3f2-247d-4d39-966e-e04a8cec55d5)

### Changes made

The code to populate the UX is the same as the IMGUI code - but split out to generate a tree before then generating the UI specific tree.

Also fixed up some styling on the control scheme info which is in the binding UI to be more consistent with the IMGUI version. Bolded and fixed margins.

### Notes

Note this version doesn't store the full tree state - just the expanded/not level.
This means I didn't need to add the static dictionary which was present in the IMGUI version.
I didn't feel this was critical and reduced static state.

I also had to make the top level expansion based on e checkbox rather than a 'fold out' / top of tree as the UI TK treeView doesn't have API to react to the expand/collapse state.
^ This comment is old - I've added a Foldout to replace the Toggle now

### Checklist

Before review:

- Changelog entry added.
- No tests added/changed
- No new doc

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

lyndon-unity committed:
Add 'matching controls' into tree view of new UI TK action asset UI

lyndon-unity committed:
Tightened up layout

Tightened up layout of control scheme text and the matching paths tree view labels
Removed the 'show matching paths' checkbox
Attempted to save and restore expansion of the 'Derived Paths' but waiting on feedback on how to respond to expand/collapse from UI TK team.

lyndon-unity committed:
Tidied up UX for matching paths

Moved some more of the style to uss file
Changed a label to a HelpBox to match UGUI version
Switched to toggle/checkbox for the main "Show matching bindings" part as we can't detect expand/collapse in tree
Removed an unused callback
Renamed some functions to be clearer

lyndon-unity committed:
Updated change log and fixed formatting

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/BindingPropertiesView.cs:
```diff
@@ -64,10 +69,261 @@ public override void RedrawUI(ViewState viewState)
 
+        private void ShowDerivedBindings(ViewState viewState)
+        {
+        }
```
Is this method unused?

lyndon-unity replied:
Yes I think I forgot to remove - will do that now.

lyndon-unity replied:
Removed now.

ritamerkl reviewed (commented):
Good to see this working! 
I just have one major point- I think it could be worth outsourcing the biggest part of the DrawMatchingControlPath and connected methods to an external script (or other related existing classes). Like that we could keep the PropertiesView clean and "logic free". 
I also added one minor comment on a potential unused method to remove.

lyndon-unity committed:
Removed unused function

lyndon-unity commented:
> Good to see this working! I just have one major point- I think it could be worth outsourcing the biggest part of the DrawMatchingControlPath and connected methods to an external script (or other related existing classes). Like that we could keep the PropertiesView clean and "logic free". I also added one minor comment on a potential unused method to remove.

Yes I agree I can take a look at spitting those out

lyndon-unity committed:
Split out MatchingControlPaths to separate source file

ritamerkl reviewed (approved):
LGTM now

Pauliusd01 commented:
Is the style used wrong? Why is the derived binding text selectable? It wasn't like that in the IMGUI version and I don't see why that should be changed

Pauliusd01 commented:
https://github.com/Unity-Technologies/InputSystem/assets/54306142/d77c6a20-4fc2-43e3-b976-4d44f3b3cc78

Since this is a toggle now I think its state should be saved, seems annoying to have to re-enable it between window restarts

Pauliusd01 reviewed (changes requested):
Just minor notes from me. Functionality seems the same and works without issues

lyndon-unity committed:
Can no longer select items in the matching controls tree

So like IMGUI version

lyndon-unity commented:
> Is the style used wrong? Why is the derived binding text selectable? It wasn't like that in the IMGUI version and I don't see why that should be changed

I've fixed that - it was just the default for the UI TK tree view.
I've set it to none to match IMGUI version

lyndon-unity commented:
> I think its state should be saved

I'd like to do that but I don't have a quick fix for that so might land the update so far and look at that later.

Pauliusd01 commented:
> > I think its state should be saved
> 
> I'd like to do that but I don't have a quick fix for that so might land the update so far and look at that later.

Does it have to be a toggle then? If a foldout would solve this issue I'd prefer that, and later on we can switch it to a toggle once state saving is there

lyndon-unity committed:
Maintaining state for showing derived binding over editor close and reopen

lyndon-unity committed:
Switched from checkbox to fold out

lyndon-unity commented:
> > > I think its state should be saved
> > 
> > 
> > I'd like to do that but I don't have a quick fix for that so might land the update so far and look at that later.
> 
> Does it have to be a toggle then? If a foldout would solve this issue I'd prefer that, and later on we can switch it to a toggle once state saving is there

State is now maintained over open/close to match IMGUI.
I've just switched to a foldout now too.

lyndon-unity committed:
Formatting fix and moved one class internal;

ekcoh reviewed (approved):
LGTM, is all logic in MatchingControlPaths new? If it is then its fine as-is, otherwise it might be worthwhile avoiding duplicating logic with existing feature implementation for IMGUI? It seems like MatchingControlPaths is not tied to UI framework and would be possible to cover with unit tests?

lyndon-unity commented:
> LGTM, is all logic in MatchingControlPaths new? If it is then its fine as-is, otherwise it might be worthwhile avoiding duplicating logic with existing feature implementation for IMGUI? It seems like MatchingControlPaths is not tied to UI framework and would be possible to cover with unit tests?

The MatchingControlPaths isn't all new code - its adapted from the IMGUI version. It drops some of the code for retaining the expand/collapse state of individual parts of the hierarchy as that had a static dictionary that seemed lower priority for the memory overheads. To use in IMGUI directly we would have to add back in the static dictionary and store extra state in the returned data. I decided to duplicate as the IMGUI version would be deprecated in future.
^ but I have now removed duplicated code and refactored a little so it can be shared with IMGUI and UI TK.

Tests should be added but I felt given time constraints that part could be added later.

lyndon-unity committed:
Removed some duplicated code and reused matching controls in IMGUI and UITK editors

And fixed the foldout callback to only trigger on the foldout and not its children

lyndon-unity committed:
Merge branch 'develop' into add_ui_tk_matching_controls

lyndon-unity committed:
Fixed issue when only IMGUI editor active (no project wide input actions)

lyndon-unity committed:
UI TK specific elements for project wide input actions protected with define

Pauliusd01 reviewed (approved):
LGTM, checked that the matching bindings are correct for different bindings in both ProjectWideActions and custom action assets.

lyndon-unity committed:
Formatting fix

ekcoh commented:
> > LGTM, is all logic in MatchingControlPaths new? If it is then its fine as-is, otherwise it might be worthwhile avoiding duplicating logic with existing feature implementation for IMGUI? It seems like MatchingControlPaths is not tied to UI framework and would be possible to cover with unit tests?
> 
> The MatchingControlPaths isn't all new code - its adapted from the IMGUI version. It drops some of the code for retaining the expand/collapse state of individual parts of the hierarchy as that had a static dictionary that seemed lower priority for the memory overheads. To use in IMGUI directly we would have to add back in the static dictionary and store extra state in the returned data. I decided to duplicate as the IMGUI version would be deprecated in future. ^ but I have now removed duplicated code and refactored a little so it can be shared with IMGUI and UI TK.
> 
> Tests should be added but I felt given time constraints that part could be added later.

I see, great to avoid duplicating too much. OK to postpone the test part temporarily but in that case please file a ticket on the remaining part from your perspective.

lyndon-unity committed:
Merge branch 'develop' into add_ui_tk_matching_controls
